### DISCUSSION TITLE
.NET HTTP/3 support on macOS

### DISCUSSION THREAD

**Discussion by alexandrehtrb** [dotnet/runtime#107614] [Q&A] [answered]
Hello,

Is HTTP/3 support on macOS planned for .NET? (does it already have support?)

**Comment by teo-tsirpanis** [comment]
HTTP/3 is not supported for [servers](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/http3?view=aspnetcore-8.0#macos) or [clients](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/quic/quic-overview#macos) on macOS because macOS lacks the necessary low-level APIs. See [related discussion](https://forums.developer.apple.com/forums/thread/724369).

Maybe Xamarin's [`NSUrlSessionHandler`](https://github.com/xamarin/xamarin-macios/blob/main/src/Foundation/NSUrlSessionHandler.cs) supports HTTP/3 or can support it? I'm not sure.

**Reply by 0x1d7** [reply]
Am I reading the response incorrectly from the Apple engineer who states the Secure Transport framework used by .NET does not and will never support HTTP/3, use the Network framework?

If .NET isn't using the SecureTransport framework, ignore me.

"Network framework supports TLS-over-QUIC, and that support includes TLS [1]. It also supports TLS-over-TCP and TLS-over-UDP.

If you were previously using Secure Transport for networking, one of these two frameworks is the correct path forward. If you can’t get on that path then you will need to build your own TLS implementation."

**Reply by am11** [reply]
The groundwork for switching to NetworkFramework is in progress: https://github.com/dotnet/runtime/pull/104835.

**Reply by liveans** [reply]
We're still using `SecureTransport` for `SslStream` and currently implementing Network Framework for TLS 1.3 in #104835, and not for QUIC, and there is no plan to use Network Framework QUIC APIs in the near future. 

Although, I had some experiment with QUIC APIs in Network Framework and currently their model doesn't fit into us, but yes @am11 is right, this is a introduction of Network Framework to the .NET codebase and this is good if we need to use it in the future.


Back to the original question, on the OSX QUIC and HTTP/3 is partially supported at the moment, it will require third party package manager with some configuration. See: https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/quic/quic-overview#macos

/cc @rzikm @ManickaP @wfurt

**Reply by wfurt** [reply]
Even if we switch TLS implementation our Http3 & Quick is powered by MsQuic at the moment (and foreseeable future) and it does not use SslStream or any our crypto & PAL primitives. Getting the Brew package (or build & bring your own) is probably best option at the moment @0x1d7 if you want HTTP/3 on macOS

**Reply by 0x1d7** [reply]
I'll be patient, thanks for the replies everyone!

**Comment by MihaZupan** [comment]
cc: @ManickaP @wfurt

**Comment by ManickaP** [comment]
Officially it isn't. But all you need is msquic lib to make it work. Once you place it next to the System.Net.Quic.dll, QUIC and H/3 will light up. Note that there's no official msquic package for OSX, so you'd need to build it yourself from the sources: https://github.com/microsoft/msquic.

Let me know if this is something you'd like to try, I can share more details and potentially help with any issues you hit.

**Reply by alexandrehtrb** [reply]
Are there plans for an official support?

**Reply by ManickaP** [reply, marked as answer]
Yep. As @wfurt mentioned bellow, @liveans is looking into it. But we don't have a set deadline and it might get pushed due to other higher prio work. So please don't hold your breath :smile:

**Comment by wfurt** [comment]
we have `brew` formula if you want to give it try @alexandrehtrb. @liveans is working on submitting that to brew repo.
